### PR TITLE
Optimize plate grid metadata

### PR DESCRIPTION
**Pull request by chris-allan**
## Summary

This change improves the performance of the `plateGrid_json` method called when displaying a plate in three main ways:

* getting the grid size
* retrieving wells in the plate
* generating the metadata for the plate grid

Various tests have been added to cover the expected behaviour and were used throughout development to attempt to ensure that no API deviation has taken place.

## Benchmarks

Based on some rough benchmarks, these changes represent an order of magnitude improvement in the overall request time. Bringing the overall request time down below the high watermark of 1 second for all OMERO.web calls.

The dataset used is the same as that outlined in #3695/#3753: 4608 Images (384 Wells, 12 Fields).

Before:

time to set grid constraints: 0:00:01.567579
time to get well grid: 0:00:00.012914
time to get wells in grid: 0:00:02.308659

After:

time to set grid constraints: 0:00:00.005936
time to get well grid: 0:00:00.086065
time to get wells in grid: 0:00:00.093965

### Grid Size
Previously, the grid size was calculated by first retrieving (via the query service) all wells in a plate, then iterating through the results to find the maximum row and column values. This change now uses HQL to find the maximum directly in the query. For the overall request, this represents a modest speed boost as all wells need to be retrieved eventually anyway, but combined with the other changes here the total number of Python objects created should still be reduced significantly.

### Retrieving Wells
With this change, instead of creating Python wrapper objects for every well in a plate, then using the wrapper to retrieve the necessary metadata, we now request the needed fields directly via HQL.

### Generating Grid Metadata
Combined with the change in how wells are retrieved, we now also avoid having to iterate every potential well in a plate. Instead, we iterate only the list of existing wells, and use the results of the Hibernate query to populate the relevant row/column in the 2D results array.

/cc @jballanc, @will-moore, @aleksandra-tarkowska, @knabar

### CODE REVIEW

File: components/tools/OmeroWeb/test/integration/test_plategrid.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+# All rights reserved.
+#
+# Use is subject to license terms supplied in LICENSE.txt
+
+"""
+   Integration tests for the "plategrid" module.
+"""
+
+import pytest
+import test.integration.library as lib
```
Comment (ximenesuk) [inline components/tools/OmeroWeb/test/integration/test_plategrid.py]: The library is not longer here, @sbesson relocated it.
`import library as lib`
should work now.

**Comment by chris-allan** [comment]
Thanks, @ximenesuk. All ported and running successfully:

```
callan@valeera:~/code/ome.git/components/tools/OmeroPy$ ./setup.py test -t test/integration/gatewaytest/test_image_wrapper.py 
WARNING: '' not a valid package name; please use only.-separated package names in setup.py
running test
False
running egg_info
writing target/omero_client.egg-info/PKG-INFO
writing top-level names to target/omero_client.egg-info/top_level.txt
writing dependency_links to target/omero_client.egg-info/dependency_links.txt
reading manifest file 'target/omero_client.egg-info/SOURCES.txt'
writing manifest file 'target/omero_client.egg-info/SOURCES.txt'
running build_ext
============================= test session starts ==============================
platform linux2 -- Python 2.7.8 -- py-1.4.20 -- pytest-2.5.2
collected 3 items 

test/integration/gatewaytest/test_image_wrapper.py ...

=========================== 3 passed in 1.49 seconds ===========================
```

```
callan@valeera:~/code/ome.git/components/tools/OmeroPy$ ./setup.py test -t test/integration/gatewaytest/test_plate_wrapper.py 
WARNING: '' not a valid package name; please use only.-separated package names in setup.py
running test
False
running egg_info
writing target/omero_client.egg-info/PKG-INFO
writing top-level names to target/omero_client.egg-info/top_level.txt
writing dependency_links to target/omero_client.egg-info/dependency_links.txt
reading manifest file 'target/omero_client.egg-info/SOURCES.txt'
writing manifest file 'target/omero_client.egg-info/SOURCES.txt'
running build_ext
============================= test session starts ==============================
platform linux2 -- Python 2.7.8 -- py-1.4.20 -- pytest-2.5.2
collected 1 items 

test/integration/gatewaytest/test_plate_wrapper.py .

=========================== 1 passed in 1.51 seconds ===========================
```

```
callan@valeera:~/code/ome.git/components/tools/OmeroWeb$ ./setup.py test -t test/integration/test_plategrid.py
WARNING: '' not a valid package name; please use only.-separated package names in setup.py
running test
False
running egg_info
writing OmeroWeb.egg-info/PKG-INFO
writing top-level names to OmeroWeb.egg-info/top_level.txt
writing dependency_links to OmeroWeb.egg-info/dependency_links.txt
reading manifest file 'OmeroWeb.egg-info/SOURCES.txt'
writing manifest file 'OmeroWeb.egg-info/SOURCES.txt'
running build_ext
============================= test session starts ==============================
platform linux2 -- Python 2.7.8 -- py-1.4.20 -- pytest-2.5.2
collected 8 items 

test/integration/test_plategrid.py ........

========================== 8 passed in 11.24 seconds ===========================
```

**Comment by pwalczysko** [comment]
BLOCKING BUG:

Wells do not appear in Web UI for freshly imported plates (via Insight, imported ``test_images_good/bd-pathway``) at all.
``user-11`` group read-write-1 on trout merge, Plate ID 2002

```
("error in call","Traceback (most recent call last):
  File "/opt/hudson/workspace/OMERO-5.1-merge-deploy/src/dist/lib/python/omeroweb/webgateway/views.py", line 1179, in wrap
    rv = f(request, *args, **kwargs)
  File "/opt/hudson/workspace/OMERO-5.1-merge-deploy/src/dist/lib/python/omeroweb/webgateway/views.py", line 1352, in plateGrid_json
    rv = plateGrid.metadata
  File "/opt/hudson/workspace/OMERO-5.1-merge-deploy/src/dist/lib/python/omeroweb/webgateway/plategrid.py", line 61, in metadata
    if acq_date.val is not None and acq_date.val > 0:
AttributeError: 'NoneType' object has no attribute 'val'
")
```

**Comment by pwalczysko** [comment]
(please note that in case this PR is in Mondays merge build with the https://github.com/openmicroscopy/openmicroscopy/pull/3776#issuecomment-98125122, this would seriously hamper testing workflows) @chris-allan @jburel @sbesson

**Comment by will-moore** [comment]
I'm getting the same error when I click on a Plate generated by Plate_To_Dataset script from a bunch of newly imported tiffs:
```
deploy/src/dist/lib/python/omeroweb/webgateway/plategrid.py", line 61, in metadata
    if acq_date.val is not None and acq_date.val > 0:
AttributeError: 'NoneType' object has no attribute 'val'
```

**Comment by pwalczysko** [comment]
The problem is not contained to new imports only - also previously imported plates have this problem, but not all of them, strangely.

**Comment by chris-allan** [comment]
There are two parts to this:

* A typo from @jballanc, where the incorrect variable is checked for nullability
* The change in 5.1.x where `acquisitionDate` nullability is now allowed

As `acquisitionDate` cannot be null in 5.0.x and earlier we never tested for that condition. That explains why data imported previously does not exhibit the problem. We will add a test for this condition and fix the typo.

**Comment by chris-allan** [comment]
The above two commits contain the adjusted test changes and the bug fix.

Just for future reference @jballanc, an OMERO rtype `.val` can never be `None` in Python. Ice does not support nullable primitive types. It is the containing object itself, `rtime` in this case, that provides nullability.

/cc @jballanc

**Comment by chris-allan** [comment]
After some discussion with @sbesson I have added the above two commits to both make the tests run quicker and provide containment emulation with respect to the xunit based style used exclusively in 5.1.x.

Runtime speedup is ~5x, the test run is much safer as there is no possibility of polluting the state of other tests and the usability semantics are essentially unchanged when porting from the 5.0.x `ITest` class hierarchy.

**Comment by will-moore** [comment]
@chris-allan It looks like your commit #8715c20 is not on trout today, since I'm still seeing the same error at https://trout.openmicroscopy.org/merge/webgateway/plate/2053/0/ 

```
  File "/opt/hudson/workspace/OMERO-5.1-merge-deploy/src/dist/lib/python/omeroweb/webgateway/plategrid.py", line 61, in metadata
    if acq_date.val is not None and acq_date.val > 0:
AttributeError: 'NoneType' object has no attribute 'val'
```
Don't know what the problem is there?

**Comment by chris-allan** [comment]
Bummer. Thoughts, @joshmoore?

**Comment by mtbc** [comment]
It does seem to be merged in:
```
$ git log remotes/scc/merge/develop/latest | grep 8715c20
commit 8715c200f53bb233321f403fc2f6598df765b5aa
```

**Comment by joshmoore** [comment]
My apologies. This is caused by my failing to click on "5.1-merge-deploy" after running "5.1-merge-build".

**Comment by joshmoore** [comment]
@will-moore : what was the status of this?

**Comment by will-moore** [comment]
Works fine now.

**Comment by joshmoore** [comment]
Thanks all.

**Pull request merged**